### ISSUE TITLE
[Research] Support ffcv backend

### ISSUE THREAD

**Issue by rentainhe** [Oneflow-Inc/libai#280]
## 讨论如何支持FFCV格式
主要参考:
- https://github.com/libffcv/ffcv
- https://github.com/libffcv/ffcv/issues/78

## FFCV的使用
- 首先最关键的点是需要制作FFCV的数据集: 参考官方例子，可以用下面的代码简单做一个FFCV的`CIFAR10`数据集
```python
datasets = {
    'train': torchvision.datasets.CIFAR10(root="./", train=True, download=True),
    'test': torchvision.datasets.CIFAR10('./', train=False, download=True)
}
for (name, ds) in datasets.items():
    writer = DatasetWriter(f'./cifar_{name}.beton', {
        'image': RGBImageField(),
        'label': IntField()
    })
    writer.from_indexed_dataset(ds)
```
数据集会保存为`.beton`的格式，在FFCV的[issue](https://github.com/libffcv/ffcv/issues/78)里有说到:
> One of the many reasons behind the speed behind FFCV is that it uses an optimized format to store the dataset. You will need to convert your dataset before loading it.

- 然后是使用FFCV的Loader模块
```python
from typing import List

import torch as ch
import torchvision

from ffcv.fields.decoders import IntDecoder, SimpleRGBImageDecoder
from ffcv.loader import Loader, OrderOption
from ffcv.pipeline.operation import Operation
from ffcv.transforms import RandomHorizontalFlip, Cutout, \
    RandomTranslate, Convert, ToDevice, ToTensor, ToTorchImage
from ffcv.transforms.common import Squeeze

CIFAR_MEAN = [125.307, 122.961, 113.8575]
CIFAR_STD = [51.5865, 50.847, 51.255]

BATCH_SIZE = 16

loaders = {}
for name in ['train', 'test']:
    label_pipeline: List[Operation] = [IntDecoder(), ToTensor(), ToDevice('cuda:0'), Squeeze()]
    image_pipeline: List[Operation] = [SimpleRGBImageDecoder()]

    # Add image transforms and normalization
    if name == 'train':
        image_pipeline.extend([
            RandomHorizontalFlip(),
            RandomTranslate(padding=2),
            Cutout(8, tuple(map(int, CIFAR_MEAN))), # Note Cutout is done before normalization.
        ])
    image_pipeline.extend([
        ToTensor(),
        ToDevice('cuda:0', non_blocking=True),
        ToTorchImage(),
        Convert(ch.float16),
        torchvision.transforms.Normalize(CIFAR_MEAN, CIFAR_STD),
    ])

    # Create loaders
    loaders[name] = Loader(f'./cifar_{name}.beton',
                            batch_size=BATCH_SIZE,
                            num_workers=8,
                            order=OrderOption.RANDOM,
                            drop_last=(name == 'train'),
                            pipelines={'image': image_pipeline,
                                       'label': label_pipeline})
```
对于这个`image_pipeline`有几个接口说明一下作用
- `RandomHorizontalFlip`, `RandomTranslate`, `Cutout`这几个是只适配于ffcv数据格式的数据增强操作
- `ToTensor()`转换为Torch的tensor，但是格式是`(H, W, C)`, 所以需要使用`ToTorchImage()`这个接口来将输入的tensor转换为`C, H, W`的格式，后续才能使用`torchvision`里的`Normalize`操作

## FFCV兼容的难度
- 数据增强操作只适用于ffcv特定的数据集，并且每个数据增强操作对接口都有一定的要求，不能独立出来兼容torchvision的数据增强
```python
from ffcv.transforms import RandomHorizontalFlip
transfrom_test = torchvision.transforms.Compose(
    [torchvision.transforms.Resize(32), RandomHorizontalFlip(), torchvision.transforms.ToTensor()]
)
dataset = torchvision.datasets.CIFAR10(root="./", train=True, download=True, transform=transfrom_test)
```
试图看看能不能单独使用`RandomHorizontalFlip`结果会报错，实现与torch的不一样，所以只能适配ffcv自己的数据集

- 不太好衡量兼容的性价比，因为实现的数据增强太少了，并且只有在调用完`ToTorchImage()`后才能得到一个pytorch的tensor，才能对这个tensor进行操作，数据增强太少无法满足训练需求

## FFCV可能的兼容方案
- 我能想到的只有重写几个`dataset.py`并且在得到数据后手动从pytorch的tensor转换为oneflow的tensor，再接入libai进行训练，ffcv的内置接口都无法单独拿出来使用

**Comment by rentainhe** [comment]
FFCV主要的加速看来还是在数据集和数据增强部分，但是有很明显的缺点是不考虑和原生torch的dataloader的兼容性，而且数据增强太少，不太符合训练需求，有很多新功能估计还在RoadMap上，FFCV还没开始做
